### PR TITLE
Adding safety check to Streamlet

### PR DESCRIPTION
**Pull request by alxiong**
Currently I have made the following improvements:

- add more code comments, properly indent all code blocks.
- add `id` field to `BlockType` to distinguish different block proposed in the same epoch 
  - this is necessary to express the `NoDoubleVotePerEpoch` property (as the paper suggested that node should vote for the _first_ proposal heard from the leader)
- switch `i` counter to `localEpoch` for better readability (I have tested out its correctness in [02f1597](https://github.com/kshehata/streamlet_tla/commit/02f15979df848deae02c970d5d1608db03aa2531) before merging into `Streamlet.tla` so I'm fairly confident it's equivalent to your PartialSync code, just more understandable IMHO. )
- add properties like `PartialSynchrony`, `LocalEpochCorrectness` etc. 
- added `Consistency` rule directly reflecting the non-conflicting finalized chain between any two honest replicas from Theorem 3 of the paper.

- [x] Adding safety properties regarding finalized blocks

### CODE REVIEW

**Comment by alxiong** [comment]
I discover a bug I introduced: when I increase `GST=3` (with 4 rounds), our `LocalEpochCorrectness` gets violated. (this violation didn't occur with lower GST and lower total rounds, so I miss this bug earlier) 

The essence of the issue is that before GST, our previous "timer" is awaiting the leader of that round to send a message,
`await (\E m \in messages : (m.block.epoch = currentEpoch /\ Leaders[currentEpoch] \in m.block.sigs));`
then we allow the timer to increase the round number. However, we (or I should say I) forgot to enforce other (non-leader) processes to catch up with that round first before moving to the next round, causing `currentEpoch -localEpoch > 1`.

In another word, here's simplified transcript of violation:
```tla
Leaders == <<n1, n2, n3, n1>>
GST == 3
localEpochs: n1 :> 1 @@ n2:> 1 @@ n3 :> 1

n1 sends message in Epoch 1
timer increment currentEpoch = 2
n2 Receive and update its localEpoch := 2
n2 sends message in Epoch 2
timer increment currentEpochh = 3

\* However, at this point, n1 and n3 still have localEpoch = 1 !!
```

And my fix is to
- ensure timer only moves forward if every process already caught up: `await \A r \in Nodes: localEpochs[r] = currentEpoch;`
- ensure our `Propose/Start/Send` will only happen if `localEpoch = currentEpoch`, or else we skip to `ReceiveOrSyncEpoch` to increment `localEpoch`

In short, after fixing the bug in [4d1e888](https://github.com/kshehata/streamlet_tla/pull/1/commits/4d1e88881c1b78a3c1f18e5c678cad9e69db85ca), now I tried various GST and `Leaders` length, all properties are satisfied.
